### PR TITLE
Modify xdma driver

### PR DESCRIPTION
**Pull request by wm-o**
Modify the driver

Originally a character device driver to transform it into a network device driver enabling communication with the Linux network stack

### CODE REVIEW

File: XDMA/linux-kernel/xdma/libxdma.c [hunk 1]
```diff
@@ -1375,61 +1382,133 @@ static irqreturn_t xdma_isr(int irq, void *dev_id)
...
+			spin_unlock_irqrestore(&priv->desc_lock[1], flag);
+			
+			/* Check if the other descriptor is ready */
+			desc = priv->desc[0];
+			spin_lock_irqsave(&priv->desc_lock[0], flag);
+			if (desc->src_addr_hi != 0 && desc->src_addr_lo != 0) {
+				desc1_status = DESC_READY;
+			}
+			spin_unlock_irqrestore(&priv->desc_lock[0], flag);
+		}
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/libxdma.c]: This block is almost same as the block above it.
How about create a simple function for it and call it with `index` as a parameter

File: XDMA/linux-kernel/xdma/libxdma.c [hunk 2]
```diff
@@ -1375,61 +1382,133 @@ static irqreturn_t xdma_isr(int irq, void *dev_id)
...
+			/* TX descriptor 2 */
+			spin_lock_irqsave(&priv->tx_lock, flag);
+			bus_addr = priv->bus_addr[1];
+			w = cpu_to_le32(PCI_DMA_L(bus_addr));
+			iowrite32(w, xdev->bar[1] + DESC_REG_LO);
+			w = cpu_to_le32(PCI_DMA_H(bus_addr));
+			iowrite32(w, xdev->bar[1] + DESC_REG_HI);
+			priv->last = 1;
+			iowrite32(DMA_ENGINE_START, &priv->tx_engine->regs->control);
+			spin_unlock_irqrestore(&priv->tx_lock, flag);
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/libxdma.c]: Also, this block could be extracted as a function

File: XDMA/linux-kernel/xdma/libxdma.h [hunk 3]
```diff
@@ -219,7 +219,10 @@
 
 #define MAX_NUM_ENGINES (XDMA_CHANNEL_NUM_MAX * 2)
 #define H2C_CHANNEL_OFFSET 0x1000
+
 #define SGDMA_OFFSET_FROM_CHANNEL 0x4000
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/libxdma.h:223]: Is this TX?

File: XDMA/linux-kernel/xdma/xdma_netdev.c [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+void rx_desc_set(struct xdma_desc *desc, dma_addr_t addr, u32 len)
+{
+        u32 control_field;
+        u32 control;
+
+        desc->control = cpu_to_le32(DESC_MAGIC);
+        control_field = XDMA_DESC_STOPPED;
+        control_field |= XDMA_DESC_EOP; 
+        control_field |= XDMA_DESC_COMPLETED;
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/xdma_netdev.c]:
```suggestion
        control_field = (
            XDMA_DESC_STOPPED
            | XDMA_DESC_EOP
            | XDMA_DESC_COMPLETED
        );
```

File: XDMA/linux-kernel/xdma/xdma_netdev.c [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                        return DESC_EMPTY;
+                }
+                tx_desc_set(desc, addr, len);
+                priv->skb[index] = skb;
+                spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+                return DESC_READY;
+        }
+        spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+        return DESC_BUSY;
+}
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/xdma_netdev.c]: What is different from `check_desc1_staus`?
It seems to be merged into single function and generalise to use more descriptors

File: XDMA/linux-kernel/xdma/xdma_netdev.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+#include "xdma_mod.h"
+
+#define RX_METADATA_SIZE 16
+#define TX_METADATA_SIZE 8
+
+#define DESC_REG_LO SGDMA_OFFSET_FROM_CHANNEL + 0x80
+#define DESC_REG_HI SGDMA_OFFSET_FROM_CHANNEL + 0x84
+
+#define DESC_REG_LO_RX SGDMA_OFFSET_FROM_CHANNEL_RX + 0x80
+#define DESC_REG_HI_RX SGDMA_OFFSET_FROM_CHANNEL_RX + 0x84
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/xdma_netdev.h]: What is 80 and 84?

Comment (tribela) [inline XDMA/linux-kernel/xdma/xdma_netdev.h]: Must use parenthesis when defining an expression
```suggestion
#define DESC_REG_LO (SGDMA_OFFSET_FROM_CHANNEL + 0x80)
#define DESC_REG_HI (SGDMA_OFFSET_FROM_CHANNEL + 0x84)

#define DESC_REG_LO_RX (SGDMA_OFFSET_FROM_CHANNEL_RX + 0x80)
#define DESC_REG_HI_RX (SGDMA_OFFSET_FROM_CHANNEL_RX + 0x84)
```

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 7]
```diff
@@ -2,81 +2,106 @@
...
-#include "example_ptp.h"
+#include "alinx_ptp.h"
+
+static u32 read32(void * addr) {
+        return ioread32(addr);
+}
+
+static void write32(u32 val, void * addr) {
+        iowrite32(val, addr);
+}
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c]: As it is OS specific alias, I suggest that move these to another file and include it for better portability and maintainability.

```c
#ifdef LINUX
void write32(…) { /* Linux specific implementation */ }
#else
void write32(…) { /* Zephyr implementation */ }
…
…
#endif
```

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 8]
```diff
@@ -2,81 +2,106 @@
...
-        timestamp_t current_ns = example_get_timestamp(sys_count, ptp_data->ticks_scale, ptp_data->offset);;
-        timestamp_t next_pulse_ns = current_ns - (current_ns % 1000000000) + 1000000000;
-        sysclock_t next_pulse_sysclock = ((double)(next_pulse_ns - ptp_data->offset) / ptp_data->ticks_scale);
+        current_ns = alinx_get_timestamp(sys_count, ptp_data->ticks_scale, ptp_data->offset);;
+        next_pulse_ns = current_ns - (current_ns % 1000000000) + 1000000000;
+        next_pulse_sysclock = ((double)(next_pulse_ns - ptp_data->offset) / ptp_data->ticks_scale);
+#ifdef __LIBXDMA_DEBUG__
+        pr_debug("ptp%u: %s sys_count=%llu, current_ns=%llu, next_pulse_ns=%llu, next_pulse_sysclock=%llu",
+               ptp_data->ptp_id, __func__, sys_count, current_ns, next_pulse_ns, next_pulse_sysclock);
+#endif
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c]: Make this as macro function is more comfortable

```c
#ifdef __LIBXDMA_DEBUG__
#define xdma_debug(...) pr_debug(__VA_ARGS__)
#else
#define xdma_debug(...) {}
#endif // __LIBXDMA_DEBUG__
```

```c
xdma_debug("ptp…");
```

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 9]
```diff
@@ -2,81 +2,106 @@
...
         struct ptp_device_data *ptp_data = container_of(
                                 ptp,
                                 struct ptp_device_data,
                                 ptp_info);
 
         spin_lock_irqsave(&ptp_data->lock, flags);
 
-        //ptp_read_system_prets(sts);
         clock = get_sys_clock(ptp_data->xdev);
-        //ptp_read_system_postts(sts);
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c]: `ptp_read_system_{pre,post}ts` is related to PCIe latency.

FYI:
https://github.com/torvalds/linux/blob/eb6a9339efeb6f3d2b5c86fdf2382cdc293eca2c/drivers/net/ethernet/mellanox/mlxsw/spectrum_ptp.c#L118-L120

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 10]
```diff
@@ -86,53 +111,65 @@ static int example_ptp_settime(struct ptp_clock_info *ptp, const struct timespec
         ptp_data->ticks_scale = TICKS_SCALE;
 
         sys_clock = get_sys_clock(xdev);
-        hw_timestamp = example_get_timestamp(sys_clock, ptp_data->ticks_scale, ptp_data->offset);
+        hw_timestamp = alinx_get_timestamp(sys_clock, ptp_data->ticks_scale, ptp_data->offset);
+
         ptp_data->offset = host_timestamp - hw_timestamp;
-        iowrite32(RESERVED_CYCLE, xdev->bar[0] + CYCLE_1S);
+
+        set_cycle_1s(ptp_data, RESERVED_CYCLE);
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c]: As `settime` is for jumping time. cycle_1s should not be touched.
I think that previous code was draft scratch

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 11]
```diff
@@ -154,41 +191,51 @@ static int XXX_example_ptp_adjfine(struct ptp_clock_info *ptp, long scaled_ppm)
...
+
         return 0;
 }
 
 struct ptp_clock_info ptp_clock_info_init(void) {
         struct ptp_clock_info info = {
                 .owner = THIS_MODULE,
                 .name = "ptp",
-                .max_adj = 12500000, // 12500000,
+                .max_adj = 12500000,
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c]: This value could be changed when using another board/HW changed.
Make it predefined constant value

Comment (tribela) [reply XDMA/linux-kernel/xdma/alinx_ptp.c]: And it is already defined as `RESERVED_CYCLE`

File: XDMA/linux-kernel/xdma/alinx_ptp.h [hunk 12]
```diff
@@ -1,19 +1,19 @@
...
-#define NEXT_PULSE_AT_HI (0x002c)
-#define NEXT_PULSE_AT_LO (0x0030)
-#define CYCLE_1S (0x0034)
-#define SYS_CLOCK_HI (0x0380)
-#define SYS_CLOCK_LO (0x0384)
+#define NEXT_PULSE_AT_HI 0x002c
+#define NEXT_PULSE_AT_LO 0x0030
+#define CYCLE_1S 0x0034
+#define SYS_CLOCK_HI 0x0380
+#define SYS_CLOCK_LO 0x0384
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.h]: These are register memory offset. Not real value.
How about this?
```suggestion
#define REG_NEXT_PULSE_AT_HI 0x002c
#define REG_NEXT_PULSE_AT_LO 0x0030
#define REG_CYCLE_1S 0x0034
#define REG_SYS_CLOCK_HI 0x0380
#define REG_SYS_CLOCK_LO 0x0384
```

File: XDMA/linux-kernel/xdma/alinx_ptp.h [hunk 13]
```diff
@@ -1,19 +1,19 @@
-#ifndef EXAMPLE_PTP_H
-#define EXAMPLE_PTP_H
+#ifndef ALINX_PTP_H
+#define ALINX_PTP_H
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.h:2]: How about `pragma once`? does it causes problem?

File: XDMA/linux-kernel/xdma/alinx_ptp.c [hunk 14]
```diff
@@ -211,16 +258,18 @@ struct ptp_device_data *ptp_device_init(struct device *dev, struct xdma_dev *xde
                 return NULL;
         }
 
+#ifdef __LIBXDMA_DEBUG__
+        ptp->ptp_id = ptp_cnt++;
+#endif
+
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_ptp.c:251]: What is it for?

File: XDMA/linux-kernel/xdma/alinx_ptp.h (hunk 13 above)
Comment (Lahyllas) [reply XDMA/linux-kernel/xdma/alinx_ptp.h:2]: I think it won't cause any problem but other header files in this repository follows this style. Shall I change all of them?

File: XDMA/linux-kernel/xdma/alinx_ptp.c (hunk 14 above)
Comment (Lahyllas) [reply XDMA/linux-kernel/xdma/alinx_ptp.c:251]: This is used to distinguish which ptp is generating the log.
``xdma_debug("ptp%u: %s cycle_1s=%u", ptp_data->ptp_id, __func__, cycle_1s);``

For example, when the CYCLE_1S register of ptp1 is written, the log would be like this:

``ptp1: set_cycle_1s cycle_1s=125000000``

It was very useful for me so I left it. Should I remove this?

File: XDMA/linux-kernel/xdma/libxdma.c [hunk 15]
```diff
@@ -1375,61 +1377,68 @@ static irqreturn_t xdma_isr(int irq, void *dev_id)
...
-			    (engine->magic == MAGIC_ENGINE)) {
-				mask &= ~engine->irq_bitmask;
-				dbg_tfr("schedule_work, %s.\n", engine->name);
-				schedule_work(&engine->work);
-			}
-		}
-	}
+		struct xdma_private *priv = netdev_priv(ndev);
+		struct xdma_engine *engine = &xdev->engine_h2c[0];
+
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/libxdma.c]:
```suggestion

```
Could you run a lint like clang-format?
There are some trailing spaces

File: XDMA/linux-kernel/xdma/alinx_arch.c [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+#include <zephyr/sys/sys_io.h>
+
+u32 read32(void * addr) {
+        return sys_read32((mem_addr_t)addr);
+}
+
+void write32(u32 val, void * addr) {
+        sys_write32(val, (mem_addr_t)addr);
+}
+
```
Comment (tribela) [inline XDMA/linux-kernel/xdma/alinx_arch.c:26]: I didn't expected. 👍🏼

**Review by tribela — APPROVED** [review]

File: XDMA/linux-kernel/xdma/alinx_ptp.c (hunk 14 above)
Comment (tribela) [reply XDMA/linux-kernel/xdma/alinx_ptp.c:251]: Understood. No need to remove

File: XDMA/linux-kernel/xdma/alinx_ptp.h (hunk 13 above)
Comment (tribela) [reply XDMA/linux-kernel/xdma/alinx_ptp.h:2]: Depend on your choice. "Should change all of them?" — No

**Pull request merged**